polkadot/node/network/collator-protocol/src/validator_side/mod.rs

```diff
@@ -924,6 +1023,84 @@ async fn process_incoming_peer_message<Context>(
 	}
 }
 
+// Holds off an AssetHub collation, if it's not from an invulnerable AssetHub collator. Returns
+// `true` if the collation was held off and `false` otherwise.
+fn hold_off_asset_hub_collation_if_needed(
+	state: &mut State,
+	peer_id: PeerId,
+	collator_id: &CollatorId,
+	relay_parent: Hash,
+	prospective_candidate: Option<(CandidateHash, Hash)>,
+) -> bool {
+	// If we don't know the peer we should reject the advertisement but to avoid verbosity and
+	// copy-pasted logic we'll just return `false` and let the caller handle it.
+	let maybe_para_id = state.peer_data.get(&peer_id).and_then(|pd| pd.collating_para());
+	let hold_off_duration = state.hold_off_duration;
+	let collator_id = collator_id.clone();
+	let peer_is_invulnerable = state.ah_invulnerables.contains(&peer_id);
+	let invulnerables_set_is_empty = state.ah_invulnerables.is_empty();
+
+	if maybe_para_id == Some(ASSET_HUB_PARA_ID) &&
```

Comment:
I like patterns presented in https://github.com/zakirullin/cognitive-load - which I've used for many years before recently seeing that repo, but that repo presents the patterns well! I think you too like these patterns :). But I found room for small improvement here:
* return early(earlier)
* named intermediary variables

So I'd change (I think `needs_hold_off` is a reasonable name?):

```rust
let needs_hold_off = maybe_para_id == Some(ASSET_HUB_PARA_ID) && !peer_is_invulnerable && 	!invulnerables_set_is_empty;

if !needs_hold_off {
	gum::trace!(
		target: LOG_TARGET,
		?maybe_para_id,
		peer_is_invulnerable,
		invulnerables_set_is_empty,
		?peer_id,
		?relay_parent,
		?prospective_candidate,
		"Collation not held off",
	);

	return false
}

// Need to hold off...
CODE TO HOLD OFF HERE
```

Reply:
I agree this is more readable and usually I use this rule too :)
I got a bit carried away with this`if` though :) I'll fix it.